libevmasm/Inliner.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * @file Inliner.cpp
+ * Inlines small code snippets by replacing JUMP with a copy of the code jumped to.
+ */
+
+#include <libevmasm/Inliner.h>
+
+#include <libevmasm/AssemblyItem.h>
+#include <libevmasm/SemanticInformation.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/slice.hpp>
+
+#include <optional>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::evmasm;
+
+bool Inliner::isInlineCandidate(u256 const& _tag, InlinableBlock const& _block) const
+{
+	assertThrow(_block.items.size() > 0, OptimizerException, "");
+
+	// Never inline tags that reference themselves.
+	for (AssemblyItem const& item: _block.items)
+		if (item.type() == PushTag)
+			if (_tag == item.data())
+				return false;
+
+	return true;
+}
+
+map<u256, Inliner::InlinableBlock> Inliner::determineInlinableBlocks(AssemblyItems const& _items) const
+{
+	std::map<u256, ranges::span<AssemblyItem const>> inlinableBlockItems;
+	std::map<u256, uint64_t> numPushTags;
+	std::optional<size_t> lastTag;
+	for (auto&& [index, item]: _items | ranges::views::enumerate)
+	{
+		// The number of PushTag's approximates the number of calls to a block.
+		if (item.type() == PushTag)
+			numPushTags[item.data()]++;
+
+		// We can only inline blocks with straight control flow that end in a jump.
+		if (lastTag && SemanticInformation::breaksCSEAnalysisBlock(item, true))
```

Comment:
Actually, the `_msizeImportant` flag can be safely set to `false`, can't it? If we set it to `false`, we can also inline calls to `keccak` and `mload`.